### ISSUE TITLE
Some bug in simpleServer.py

### ISSUE THREAD

**Issue by yoid2000** [gda-score/anonymization-mechanisms#2]
@reinhardmunz 

Some bug has cropped up in the code that an intern wrote a couple years ago. It had been working, so I'm not sure what has changed where. 

The code `simpleServer()` at uber/examples is a flask application running at `https://db001.gda-score.org/ubertool`

Trying to test it using `simpleClientFQDN()` (from an anaconda cmd window on a windows machine) times out at the client.

The log at the server (which I get from Pascal) shows this:

```
Oct 26 11:23:33 srv-53-01 flask[4214]: JSON sent by Client: {u'query': u"Select count(*) from transactions where operation = 'VKLAD' ", u'sid': u'', u'budget': u'6.0', u'dbname': u'raw_banking', u'[2020-10-26 11:23:33,360] ERROR in app: E
Oct 26 11:23:33 srv-53-01 flask[4214]: Traceback (most recent call last):
Oct 26 11:23:33 srv-53-01 flask[4214]:   File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1813, in full_dispatch_request
Oct 26 11:23:33 srv-53-01 flask[4214]:     rv = self.dispatch_request()
Oct 26 11:23:33 srv-53-01 flask[4214]:   File "/usr/local/lib/python2.7/dist-packages/flask/app.py", line 1799, in dispatch_request
Oct 26 11:23:33 srv-53-01 flask[4214]:     return self.view_functions[rule.endpoint](**req.view_args)
Oct 26 11:23:33 srv-53-01 flask[4214]:   File "/usr/local/lib/python2.7/dist-packages/flask_restful/__init__.py", line 458, in wrapper
Oct 26 11:23:33 srv-53-01 flask[4214]:     resp = resource(*args, **kwargs)
Oct 26 11:23:33 srv-53-01 flask[4214]:   File "/usr/local/lib/python2.7/dist-packages/flask/views.py", line 88, in view
Oct 26 11:23:33 srv-53-01 flask[4214]:     return self.dispatch_request(*args, **kwargs)
Oct 26 11:23:33 srv-53-01 flask[4214]:   File "/usr/local/lib/python2.7/dist-packages/flask_restful/__init__.py", line 573, in dispatch_request
Oct 26 11:23:33 srv-53-01 flask[4214]:     resp = meth(*args, **kwargs)
Oct 26 11:23:33 srv-53-01 flask[4214]:   File "/home/flask/flaskapp/simpleServer.py", line 108, in get
Oct 26 11:23:33 srv-53-01 flask[4214]:     query = client_request['query']  # Query
Oct 26 11:23:33 srv-53-01 flask[4214]: TypeError: 'NoneType' object has no attribute '__getitem__'
Oct 26 11:23:33 srv-53-01 flask[4214]: 127.0.0.1 - - [26/Oct/2020 11:23:33] "GET /data HTTP/1.1" 500 -
```

The error message `TypeError: 'NoneType' object has no attribute '__getitem__'` looks to me like `client_request` is set to `None` at line 108. Which is strange because just prior to that the contents of `client_request` is being printed. Maybe I'm interpreting the error message incorrectly.

In any event, could you dig into this and see what is going on? Would not hurt to make the code a bit more robust in the process...

Maybe helpful if Pascal gives you the credentials to modify / restart simpleServer.py on db001.

**Comment by reinhardmunz** [comment]
The log message "JSON sent by Client ..." is misleading. It is printing the client REQUEST there before parsing to JSON. The problem as one can see in your log message is, that you are not actually sending valid JSON in your REQUEST. It is cut off and there is part of an error message in the end. The JSON parser then returns None, which leads to the final error.

I updated the code to check this case and also correctly handle any other errors. The server should always provide a valid JSON response now, potentially with an "Error" field in the "Server Response" instead of the expected other fields there.

The program needs at least python 3.6 now, as I used f-strings. I made changes to the systemd service definitions accordingly.

Please reopen the issue if you want me to further clean/comment that particular code.

**Comment by yoid2000** [comment]
Thanks @reinhardmunz .

I gather you had the server updated at db001, because I'm no longer getting timeout error?

But now the problem seems to be that the wrong answer is being returned. For example, from `simpleClientFQDN.py`, part of the output is this:

```
{   'budget': '1.0',
    'dbname': 'raw_banking',
    'epsilon': '2.0',
    'query': 'Select count(*) from accounts',
    'sid': 281294948}
{'Server Response': {'Result': '5', 'Session ID': 281294948}}
```

The result of 5 is much to low, even accounting for noise. Furthermore, the result is the same with each repetition, suggesting that some other value is being put into the `Result` field.

Could you have a look?

**Comment by reinhardmunz** [comment]
Sorry, my bad. I introduced a bug that only returned the first character of a result. I fixed it.

I am not sure whether the app has a more basic problem in its setup though.

I think as soon as multiple concurrent requests are coming in, there is a race condition and results may be returned to any of the concurrent clients at random.

The flask-server app takes a client's request and writes it to a file in a request folder.
For any change to a file in the request folder, the uber app calculates the result and writes it to a new file in a result folder.
Meanwhile the flask-server app watches the result folder and as soon as the number of files increases, it takes the file with the last ctime and returns its contents to the client.

When multiple clients do this procedure concurrently, there is a race condition between the different result files and which flask thread reads which file first. In theory any client may receive the result of any other client at random.

Please let me know if I should look into this and change the basic setup of both apps to ensure correct mapping of results in case of concurrent requests.

**Comment by yoid2000** [comment]
I wasn't aware of that problem. Good catch. Yes, please go ahead and fix. But do the easiest thing. For instance, if it is substantially easier to just queue concurrent requests, rather than running them in parallel, then please do so. I'm not sure why that would be the case, I'm just saying we don't need to be too fancy here.

**Comment by reinhardmunz** [comment]
The problem isn't flask, it is the Uber Tool.

flask already runs in non-production, single threaded debug mode. It does exactly what you suggest, it queues requests and runs them one after another.

The Uber Tool, however, crashes on invalid inputs, e.g, if epsilon is 0.0 for a non-empty query. Such queries may happen due to bugs or once clients time out. Timeouts happened for me once I tested concurrent requests with five clients (with an intentionally low timeout value of 15s).

I changed the server to reject invalid queries. I further don't understand why the first query had to be empty with epsilon 0.0. I changed that. Any query without sid just gets a new sid. Subsequent queries with the same sid are handled as before. Empty queries with 0.0 epsilon continue to be allowed. They simply return an empty result and are otherwise regular queries. Thus, the previous behavior (=first query is empty with epsilon 0.0) continues to be valid behavior.

I modified the client and removed its empty first query.

**Comment by yoid2000** [comment]
I know that the uber interface is ugly, but please keep it as is because it
is going to be hidden behind an interface in gda-score. In fact, the
gda-score code is already written, so probably easier to put the uber
interface as it was rather than change the gda-score interface again.

your other changes seem fine to me.

PF

**Comment by reinhardmunz** [comment]
I haven't touched the uber interface (in src folder), only the simpleServer (in examples folder). I mitigated the shortcomings of the uber interface with additional checks in simpleServer. I did not fix the crash on non-empty query with 0.0 epsilon in the uber interface. I only extended the interface of simpleServer. Anything that worked before should still work.

From your answer I take that I did the correct thing by not touching the uber interface.
